cogs/utils/arg_converter.py

```diff
@@ -0,0 +1,150 @@
+        (as above)
+    @commands.command()
+    async def example(self, ctx, *args):
+      arg_converter = ArgConverter({"member": (MemberConverter(), None),
+                                    "channel": (TextChannelConverter(), )
+                                    "string_A": (FromConverter(), "hi"))
+    try:
+        args_dict = await arg_converter.convert(ctx, args)
+    except commands.BadArgument as err:
+        await ctx.send(err)
+        return
+    await ctx.send(f"{str(member)}, {channel}, {string_A}")
+
+    Here are example inputs and what is sent in the server:
+    `?example #channel`: None, #channel, hi
+    `?example aword @someone #channel`: someone#1234, #channel, word
+    `?example aword @someone"`:
+    Invalid input: Missing required argument "channel"
+    `?example #channel aword @someone woo`:
+    Invalid input: Argument "woo" could not be converted
+    """
+    def __init__(self, converters_dict):
+        self.converters_dict = converters_dict
+        # get the list of variables that are required
+        self.req_vars = [
+            key for key, value in converters_dict.items() if len(value) == 1
+        ]
+
+    async def convert(self, ctx, arguments):
+        """
+        Match the arguments to the converters. If a converter is used,
+        doesn't try to match it again. All required arguments must be
+        given, and all arguments must be able to be converted.
+
+        Order doesn't matter, except if there are multiple converters that
+        match with the same argument, in which case the first one in the
+        dictionary is used. This could lead to unexpected results:
+        for example, it doesn't matter if this is to catch two discord users,
+        but if there was a converter meant to catch any string, then a
+        converter to catch an uppercase string, then the arguments
+        ["HI", "hello"] will raise an exception, since the "any" string
+        converter is the one used with "HI", and then the only remaining
+        converter is the uppercase one, and so "hello" is not converted.
+
+        Returns a dictionary with the variables as keys, and their values.
+        """
+        # initialize the converted dict with None values
+        converted_arguments_dict = dict.fromkeys(self.converters_dict)
+        # get a copy of the list of variables that are required
+        # if there are still values in this at the end, an error will be raised
+        remaining_req_vars = self.req_vars.copy()
+        for arg in arguments:
+            # keep track if the argument is converted. If it is not after
+            # trying every converter, an error will be raised
+            arg_converted = False
+            for key in self.converters_dict:
+                try:
+                    # if the value for that key
+                    # is still none in the converted dict
+                    if not converted_arguments_dict[key]:
```

Comment:
same as before, use `is None` instead of `if not` since you could have a boolean converter

Reply:
Outdated in c6111c6